File: uva/100.cpp
#include <cstdio>
#include <cstring>

#define N 1000000

int C[N];

int cycle(int n)
{
	if(n == 1) return 1;
	if(n >= N)
	{
		if(n & 1) return 1 + cycle(3 * n + 1);
		return 1 + cycle(n >> 1);
	}

	if(!C[n])
	{
		if(n & 1) C[n] = 1 + cycle(3 * n + 1);
		else C[n] = 1 + cycle(n >> 1);
	}

	return C[n];
}

int max_cycle(int i, int j)
{
	int c, m = 0;

	while(i <= j)
	{
		c = cycle(i++);
		if(c > m) m = c;
	}

	return m;
}

int main()
{
	int a, b, c;
	memset(C, 0, sizeof(C));
	
	while(scanf("%d %d", &a, &b) != EOF)
	{
		if(a < b) c = max_cycle(a, b);
		else c = max_cycle(b, a);
		printf("%d %d %d\n", a, b, c);
	}

	return 0;
}

File: uva/100.py
import sys

N = 1000000
C = [0] * N

def cycle(n):
	if(n == 1):
		return 1
	if(n >= N):
		if(n % 2):
			return 1 + cycle(3 * n + 1)
		else:
			return 1 + cycle(n / 2)

	if(C[int(n)] == 0):
		if(n % 2):
			C[int(n)] = 1 + cycle(3 * n + 1)
		else:
			C[int(n)] = 1 + cycle(n / 2)

	return C[int(n)]

def max_cycle(i, j):
	m = 0
	while(i <= j):
		c = cycle(i)
		if(c > m):
			m = c
		i += 1
	return m

for line in sys.stdin:
	a,b = line.split()
	a = int(a)
	b = int(b)
	if(a < b):
		c = max_cycle(a, b)
	else:
		c = max_cycle(b, a)

	print(a, b, c)

